lib/Server.js

```diff
@@ -741,23 +743,102 @@ class Server {
+    const setupCallback = () => {
       this.createSocketServer();
 
       if (this.options.bonjour) {
         runBonjour(this.options);
       }
 
       this.showStatus();
+    };
 
-      if (fn) {
+    // between setupCallback and userCallback should be any other needed handling,
+    // specifically so that things are done in the right order to prevent
+    // backwards compatability issues
+    let userCallbackCalled = false;
+    const userCallback = (err) => {
+      if (fn && !userCallbackCalled) {
+        userCallbackCalled = true;
         fn.call(this.listeningApp, err);
       }
+    };
 
+    const onListeningCallback = () => {
       if (typeof this.options.onListening === 'function') {
         this.options.onListening(this);
       }
-    });
+    };
+
+    const fullCallback = (err) => {
+      setupCallback();
+      userCallback(err);
+      onListeningCallback();
+    };
+
+    // try to follow the Node standard in terms of deciding
+    // whether this is a socket or a port that we will listen on:
+    // https://github.com/nodejs/node/blob/64219741218aa87e259cf8257596073b8e747f0a/lib/net.js#L196
+    if (typeof port === 'string' && tryParseInt(port) === null) {
+      // in this case the "port" argument is actually a socket path
+      const socket = port;
+      // set this so that status helper can identify how the project is being run correctly
+      this.options.socket = socket;
+
+      const chmodSocket = (done) => {
+        // chmod 666 (rw rw rw) - octal
+        const READ_WRITE = 438;
+        fs.chmod(socket, READ_WRITE, done);
+      };
+
+      const startSocket = () => {
+        this.listeningApp.on('error', (err) => {
+          userCallback(err);
+        });
+
+        this.listeningApp.listen(socket, hostname, (err) => {
+          if (err) {
+            fullCallback(err);
+          } else {
+            chmodSocket(fullCallback);
+          }
+        });
+      };
+
+      fs.access(socket, fs.constants.F_OK, (err) => {
+        if (err) {
+          // file does not exist
+          startSocket();
+        } else {
+          // file exists
+
+          const clientSocket = new net.Socket();
+
+          clientSocket.on('error', (err) => {
+            if (err.code === 'ECONNREFUSED' || err.code === 'ENOTSOCK') {
+              // No other server listening on this socket so it can be safely removed
+              fs.unlinkSync(socket);
+
+              startSocket();
+            }
+          });
+
+          clientSocket.connect({ path: socket }, () => {
+            // if a client socket successfully connects to the given socket path,
+            // it means that the socket is in use
+            const err = new Error('This socket is already used');
+            clientSocket.destroy();
+            // do not call onListening or the setup method, since the server
+            // cannot start listening on a used socket
+            userCallback(err);
+          });
+        }
+      });
```

Comment:
Can we refactor this on utils? Maybe multiple utils

Reply:
I'll try